app/demo-geant-integration/HepMC3Reader.cc

```diff
@@ -0,0 +1,133 @@
+//---------------------------------------------------------------------------//
+#include "HepMC3Reader.hh"
+
+#include <mutex>
+#include <G4SystemOfUnits.hh>
+#include <G4TransportationManager.hh>
+#include <HepMC3/ReaderFactory.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/Constants.hh"
+
+#include "GlobalSetup.hh"
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Return non-owning pointer to a singleton.
+ */
+HepMC3Reader* HepMC3Reader::instance()
+{
+    static HepMC3Reader hepmc3_reader_singleton;
+    return &hepmc3_reader_singleton;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add HepMC3 primaries to a Geant4 event. This function is called by
+ * `G4VUserPrimaryGeneratorAction::GeneratePrimaries`.
+ *
+ * \note
+ * Current implementation is compatible with our utils/hepmc3-generator
+ * (https://github.com/celeritas-project/utils) output files, including the
+ * translated CMS' Pythia HEPEVT output files to HepMC3 format. Nevertheless,
+ * these files do not have more complex topologies with multiple vertices with
+ * mother/daughter particles. If more complex inputs are used, this will have
+ * to be updated.
+ */
+void HepMC3Reader::GeneratePrimaryVertex(G4Event* g4_event)
+{
+    static std::mutex           hepmc3_mutex;
+    std::lock_guard<std::mutex> scoped_lock{hepmc3_mutex};
+
+    HepMC3::GenEvent gen_event;
+    input_file_->read_event(gen_event);
+    CELER_ASSERT(!input_file_->failed());
+
+    CELER_LOG_LOCAL(status)
+        << "Reading HepMC3 event " << gen_event.event_number();
+
+    gen_event.set_units(HepMC3::Units::MEV, HepMC3::Units::MM); // Geant4 units
+    const auto& event_pos = gen_event.event_pos();
+
+    // Verify if vertex is inside the world volume
+    const auto inside_status
+        = G4TransportationManager::GetTransportationManager()
+              ->GetNavigatorForTracking()
+              ->GetWorldVolume()
+              ->GetLogicalVolume()
+              ->GetSolid()
+              ->Inside(
+                  G4ThreeVector{event_pos.x(), event_pos.y(), event_pos.z()});
+
+    CELER_ASSERT(inside_status == EInside::kInside);
+
+    // Add primaries to event
+    const auto& gen_particles = gen_event.particles();
+    for (const auto& gen_particle : gen_particles)
+    {
+        const auto& part_data = gen_particle->data();
+
+        if (part_data.status != 1)
+        {
+            // Skip particles that should not be tracked
+            // Status codes (page 13):
+            // http://hepmc.web.cern.ch/hepmc/releases/HepMC2_user_manual.pdf
+            continue;
+        }
+
+        auto g4_vtx = std::make_unique<G4PrimaryVertex>(
+            event_pos.x(),
+            event_pos.y(),
+            event_pos.z(),
+            event_pos.t() * CLHEP::s / celeritas::constants::c_light); // [s]
```

Comment:
To be consistent, use the `CLHEP` definition of `c` here (since it uses mm and ours is cm!). And actually does the `G4PrimaryVertex` take time in seconds, or in `s*c`? It may not need a conversion at all.

Reply:
Good point. I assumed the vertex would use standard G4 units (mm, ns), so not a fraction of `c`, but I'll double check this.